app/utils/ab_test.js

```diff
@@ -0,0 +1,80 @@
+'use strict'
+
+const cookie = require('./cookies')
+const random = require('./random')
+const crypto = require('crypto')
+const logger = require('winston')
+
+exports.uniformlyGeneratedRandomNumber = function () {
+  // AFAIK node's Math.random function is not uniformly random
+  // Hence the following dance
+  //
+  // 1. Generate a random number
+  // 2. Hash it so it is uniformly distributed
+  // 3. Modulo 100
+  let abUuid = random.randomUuid()
+  let hash = crypto.createHash('sha256')
+
+  hash.update(abUuid)
+
+  // We will Math.ceil our number so we don't have to check zeros
+  let uniformHash = hash.digest('hex')
+  let uniformNumber = parseInt(uniformHash, 16)
+  let uniformModNumber = Math.ceil(uniformNumber % 100)
+
+  return uniformModNumber
+}
+
+exports._getSession = function (req) {
+  let sessionVal = cookie.getSessionVariable(req, 'abTestId')
+  return sessionVal || null
+}
+
+exports._setSession = function (req) {
+  return cookie.setSessionVariable(
+    req,
+    'abTestId',
+    exports.uniformlyGeneratedRandomNumber()
+  )
+}
+
+exports._getOrSetSession = function (req) {
+  if (exports._getSession(req) === null) {
+    logger.info('Could not find user ab testing session, creating one')
+    exports._setSession(req)
+  }
+  return exports._getSession(req)
+}
+
+exports.switch = function (opts) {
+  let threshold = Math.floor(opts.threshold || 90)
+  let defaultVariant = opts.defaultVariant
+  let testingVariant = opts.testingVariant
+
+  return function (req, res) {
+    let sessionValue = exports._getOrSetSession(req)
+
+    let showDefault = true
```

Comment:
why is this explicitly true? why this variable?

Reply:
Is it just extra safe or is it silly? How should I refactor?

Reply:
The behaviour here is

- First calculate whether normally we would see the default
- Then if the showAbTest query param is truthy, force the ab test
- Then if the hideAbTest query param is truthy, force the default

This is superfluous, I guess we want a single query parameter to force/unforce

```
const showDefault = req.query.abTest
  ? req.query.abTest != 'yes'
  : sessionValue <= threshold
```